bookkeeper-server/src/main/java/org/apache/bookkeeper/client/LedgerCreateOp.java

```diff
@@ -200,4 +209,220 @@ private void createComplete(int rc, LedgerHandle lh) {
+            }
+            return advBuilder;
+        }
+
+        private boolean validate(CreateCallback cb, Object ctx) {
+            if (builderWriteQuorumSize > builderEnsembleSize) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderAckQuorumSize > builderWriteQuorumSize) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderAckQuorumSize <= 0) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderPassword == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderDigestType == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            if (builderCustomMetadata == null) {
+                cb.createComplete(BKException.Code.IncorrectParameterException, null, ctx);
+                return false;
+            }
+
+            return true;
+        }
+
+        @Override
+        public CompletableFuture<WriteHandle> execute() {
+            CompletableFuture<WriteHandle> counter = new CompletableFuture<>();
+            create(new SyncCreateCallback(), counter);
+            return counter;
+        }
+
+        private void create(CreateCallback cb, Object ctx) {
+            if (!validate(cb, ctx)) {
+                return;
+            }
+            LedgerCreateOp op = new LedgerCreateOp(bk, builderEnsembleSize,
+                builderWriteQuorumSize, builderAckQuorumSize, DigestType.fromApiDigestType(builderDigestType),
+                builderPassword, cb, ctx, builderCustomMetadata);
+            ReentrantReadWriteLock closeLock = bk.getCloseLock();
+            closeLock.readLock().lock();
+            try {
+                if (bk.isClosed()) {
+                    cb.createComplete(BKException.Code.ClientClosedException, null, ctx);
+                    return;
+                }
+                op.initiate();
+            } finally {
+                closeLock.readLock().unlock();
+            }
+        }
+    }
+
+    private static class CreateAdvBuilderImpl implements CreateAdvBuilder {
+
+        private Long builderLedgerId;
+        private final int builderEnsembleSize;
+        private final int builderAckQuorumSize;
+        private final int builderWriteQuorumSize;
+        private final byte[] builderPassword;
+        private final Map<String, byte[]> builderCustomMetadata;
+        private final org.apache.bookkeeper.client.api.DigestType builderDigestType;
+        private final BookKeeper bk;
+
+         private CreateAdvBuilderImpl(CreateBuilderImpl other) {
```

Comment:
hmm..I'm not convinced with how CreateBuilderImpl and CreateAdvBuilderImpl are created/implemented, now if someone wants to use createadv, they have to first create CreateBuilderImpl, then call makeAdv and then call withLedgerId method. Doesn't look correct.

Reply:
withLedgerId is options.
The API is following a fluent style and it looks very good to me, we already discussed it in the BP-15.

try (BookKeeper bk = BookKeeper.newBuilder(config).build();
      WriteAdvHandle lh = result(bk
                 .newCreateLedgerOp()
                 .withEnsembleSize(3)
                 .with(xxx)
                 .makeAdv()
                 .execute())) {
      lh.write(1,data);
}

or 
try (BookKeeper bk = BookKeeper.newBuilder(config).build();
      WriteHandle lh = result(bk
                 .newCreateLedgerOp()
                 .withEnsembleSize(3)
                 .with(xxx)               
                 .execute())) {
      lh.append(data);
}